oxidation/crates/parsec_client_types/src/local_manifest.rs

```diff
@@ -111,6 +183,99 @@ impl From<LocalFileManifest> for LocalFileManifestData {
 
 impl_local_manifest_dump_load!(LocalFileManifest);
 
+impl LocalFileManifest {
+    pub fn get_chunks(&self, block: usize) -> Vec<Chunk> {
+        self.blocks.get(block).cloned().unwrap_or_default()
+    }
+
+    pub fn is_reshaped(&self) -> bool {
+        for chunks in self.blocks.iter() {
+            if chunks.len() != 1 || !chunks[0].is_block() {
+                return false;
+            }
+        }
+        true
+    }
+
+    pub fn assert_integrity(&self) {
+        let mut current = 0;
+        for (i, chunks) in self.blocks.iter().enumerate() {
+            assert_eq!(i as u64 * *self.blocksize, current);
+            assert!(!chunks.is_empty());
+            for chunk in chunks {
+                assert_eq!(chunk.start, current);
+                assert!(chunk.start < chunk.stop.into());
+                assert!(chunk.raw_offset <= chunk.start);
+                assert!(u64::from(chunk.stop) <= chunk.raw_offset + u64::from(chunk.raw_size));
+                current = chunk.stop.into()
+            }
+        }
+        assert_eq!(current, self.size);
+    }
+
+    pub fn from_remote(remote: FileManifest) -> Result<Self, &'static str> {
+        let base = remote.clone();
+        let blocks = remote
+            .blocks
+            .into_iter()
+            .map(Chunk::from_block_access)
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .map(|b| vec![b])
+            .collect();
+
+        Ok(Self {
+            base,
+            need_sync: false,
+            updated: remote.updated,
+            size: remote.size,
+            blocksize: remote.blocksize,
+            blocks,
+        })
+    }
+
+    pub fn to_remote(
+        &self,
+        author: DeviceID,
+        timestamp: DateTime,
+    ) -> Result<FileManifest, &'static str> {
+        self.assert_integrity();
+        assert!(self.is_reshaped());
+
+        let blocks = self
+            .blocks
+            .iter()
+            .map(|chunks| chunks[0].get_block_access())
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .flatten()
+            .collect();
+
+        Ok(FileManifest {
+            author,
+            timestamp,
+            id: self.base.id,
+            parent: self.base.parent,
+            version: self.base.version + 1,
+            created: self.base.created,
+            updated: self.updated,
+            size: self.size,
+            blocksize: self.blocksize,
+            blocks,
+        })
+    }
+
+    pub fn match_remote(&self, remote_manifest: &FileManifest) -> Result<bool, &'static str> {
```

Comment:
I think we can use `-> bool` as return value, this would make this method more pleasant to use and we don't care that the match failed because the file manifest is not reshaped.

Reply:
Done :+1: